nvflare/app_common/streamers/file_downloader.py

```diff
@@ -0,0 +1,230 @@
+        with open(obj.name, "rb") as f:
+            f.seek(received_bytes)
+            chunk = f.read(num_bytes_to_send)
+
+        self.logger.debug(f"{received_bytes=}; sending {len(chunk)} bytes")
+        return ProduceRC.OK, chunk, {_StateKey.RECEIVED_BYTES: received_bytes + len(chunk)}
+
+
+class FileDownloader:
+
+    @classmethod
+    def new_transaction(
+        cls,
+        cell: Cell,
+        timeout: float,
+        timeout_cb,
+        **cb_kwargs,
+    ):
+        """Create a new file download transaction.
+
+        Args:
+            cell: the cell for communication with recipients
+            timeout: timeout for the transaction
+            timeout_cb: CB to be called when the transaction is timed out
+            **cb_kwargs: args to be passed to the CB
+
+        Returns: transaction id
+
+        The timeout_cb must follow this signature:
+
+            cb(tx_id, file_names: List[str], **cb_args)
+
+        """
+        return ObjDownloader.new_transaction(
+            cell=cell,
+            producer=_ChunkProducer(),
+            timeout=timeout,
+            timeout_cb=cls._tx_timeout,
+            app_timeout_cb=timeout_cb,
+            **cb_kwargs,
+        )
+
+    @classmethod
+    def _tx_timeout(cls, tx_id: str, objs: List[Any], app_timeout_cb, **cb_kwargs):
+        if app_timeout_cb:
+            file_names = [obj.name for obj in objs]
+            app_timeout_cb(tx_id, file_names, **cb_kwargs)
+
+    @classmethod
+    def add_file(
+        cls,
+        transaction_id: str,
+        file_name: str,
+        file_downloaded_cb=None,
+        **cb_kwargs,
+    ) -> str:
+        """Add a file to be downloaded to the specified transaction.
+
+        Args:
+            transaction_id: ID of the transaction
+            file_name: name of the file to be downloaded
+            file_downloaded_cb: CB to be called when the file is done downloading
+            **cb_kwargs: args to be passed to the CB
+
+        Returns: reference id for the file.
+
+        The file_downloaded_cb must follow this signature:
+
+            cb(ref_id: str, to_site: str, status: str, file_name: str, **cb_kwargs)
+
+        """
+        obj = _File(file_name)
+        return ObjDownloader.add_download_object(
+            transaction_id=transaction_id,
+            obj=obj,
+            obj_downloaded_cb=cls._file_downloaded,
+            app_downloaded_cb=file_downloaded_cb,
+            **cb_kwargs,
+        )
+
+    @classmethod
+    def _file_downloaded(cls, ref_id: str, to_site: str, status: str, obj: _File, app_downloaded_cb, **cb_kwargs):
+        if app_downloaded_cb:
+            app_downloaded_cb(ref_id, to_site, status, obj.name, **cb_kwargs)
+
+
+class _ChunkConsumer(Consumer):
+
+    def __init__(self, location: str):
+        Consumer.__init__(self)
+        self.location = location
+        self.file_path = os.path.join(location, str(uuid.uuid4()))
```

Comment:
if we want we could add some hash check (MD5) for file integrity in the future.

Reply:
We are downloading to a temp file path (uuid), so the caller are responsible for removing this file

Then does it matter if we should pass in "location" here?

Reply:
It's up to the caller when to delete it or deleting it at all.

The  caller can specify a "location" so that the caller can manager its downloaded files in whatever way it wants. For example, it could just download everything into the same location, and remove the location completely at the end of the job.

Reply:
Right. :-)